src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeReader.java

```diff
@@ -0,0 +1,455 @@
+        {
+            int commonPrefixLength = readCommonPrefixLength(in);
+            int compressedDimension = readCompressedDimension(in);
+            // If the compressed dimension is -1 it means that all values in the block are equal
+            if (compressedDimension == -1)
+                return visitRawDocValues(count, visitor, origIndex);
+            else
+                return visitCompressedDocValues(commonPrefixLength, in, count, visitor, origIndex);
+        }
+
+        private int readCompressedDimension(IndexInput in) throws IOException
+        {
+            int compressedDimension = in.readByte();
+            if (compressedDimension != -1 && compressedDimension != 0)
+                throw new CorruptIndexException(String.format("Dimension should be in the range [-1, 0], but was %d.", compressedDimension), in);
+            return compressedDimension;
+        }
+
+        private FixedBitSet visitCompressedDocValues(int commonPrefixLength,
+                                                     IndexInput in,
+                                                     int count,
+                                                     IntersectVisitor visitor,
+                                                     short[] origIndex) throws IOException
+        {
+            // the byte at `compressedByteOffset` is compressed using run-length compression,
+            // other suffix bytes are stored verbatim
+            int compressedByteOffset = commonPrefixLength;
+            commonPrefixLength++;
+            int i;
+
+            FixedBitSet fixedBitSet = new FixedBitSet(maxPointsInLeafNode);
+
+            for (i = 0; i < count; )
+            {
+                packedValue[compressedByteOffset] = in.readByte();
+                final int runLen = Byte.toUnsignedInt(in.readByte());
+                for (int j = 0; j < runLen; ++j)
+                {
+                    in.readBytes(packedValue, commonPrefixLength, bytesPerValue - commonPrefixLength);
+                    final int rowIDIndex = origIndex[i + j];
+                    if (visitor.visit(packedValue))
+                        fixedBitSet.set(rowIDIndex);
+                }
+                i += runLen;
+            }
+            if (i != count)
+                throw new CorruptIndexException(String.format("Expected %d sub-blocks but read %d.", count, i), in);
+
+            return fixedBitSet;
+        }
+
+        private FixedBitSet visitRawDocValues(int count, IntersectVisitor visitor, final short[] origIndex)
+        {
+            FixedBitSet fixedBitSet = new FixedBitSet(maxPointsInLeafNode);
+
+            // In our one-dimensional case, all the values in the leaf are the same, so we only
+            // need to visit once then set the bits for the relevant indexes
+            if (visitor.visit(packedValue))
+            {
+                for (int i = 0; i < count; ++i)
+                    fixedBitSet.set(origIndex[i]);
```

Comment:
Do we need to create a new bitset at all in this method? If the single value in the leaf is included, it means we could just pass a singleton "everything matches" `BitSet` right? (...or a "nothing matches" bitset otherwise?)

Reply:
Unfortunately, without rewriting this part of the Lucene code we have to do this here. We rely on the cardinality returned from the `FixedBitset` which isn't available in the `Bits` interface that helpfully implements a `MatchAllBits` and a `MatchNoBits` class.